### PR TITLE
Unify the status handle in bindingCycle

### PR DESCRIPTION
**Pull request by kerthcet**
Signed-off-by: kerthcet <kerthcet@gmail.com>

#### What type of PR is this?
/sig scheduling

#### What this PR does / why we need it:
This PR is a preparation for https://github.com/kubernetes/kubernetes/pull/112025, it helps to unify the status handle in bindingCycle so that we can de-duplicate the logic eventually.

Before, in preBind and Bind, we return success/error status but the unschedulable, now we return the status directly
and handle it in bindingCycle.

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/kubernetes/issues/103853

#### Special notes for your reviewer:
We'll leave the refactor in https://github.com/kubernetes/kubernetes/pull/112025

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/framework/interface.go [hunk 1]
```diff
@@ -250,6 +250,9 @@ func NewStatus(code Code, reasons ...string) *Status {
 
 // AsStatus wraps an error in a Status.
 func AsStatus(err error) *Status {
+	if err == nil {
```
Comment (chendave) [inline pkg/scheduler/framework/interface.go:253]: why this check is needed?

File: pkg/scheduler/framework/runtime/framework.go [hunk 2]
```diff
@@ -1030,15 +1035,20 @@ func (f *frameworkImpl) RunBindPlugins(ctx context.Context, state *framework.Cyc
 	if len(f.bindPlugins) == 0 {
 		return framework.NewStatus(framework.Skip, "")
 	}
-	for _, bp := range f.bindPlugins {
-		status = f.runBindPlugin(ctx, bp, state, pod, nodeName)
+	for _, pl := range f.bindPlugins {
```
Comment (chendave) [inline pkg/scheduler/framework/runtime/framework.go:1038]: let's revert this back. I don't think it's needed.

File: pkg/scheduler/framework/runtime/framework_test.go [hunk 3]
```diff
@@ -1704,7 +1709,7 @@ func TestPreBindPlugins(t *testing.T) {
 					inj:  injectedResult{PreBindStatus: int(framework.UnschedulableAndUnresolvable)},
 				},
 			},
-			wantStatus: framework.AsStatus(fmt.Errorf(`running PreBind plugin "TestPlugin": %w`, errInjectedStatus)),
+			wantStatus: framework.NewStatus(framework.UnschedulableAndUnresolvable, injectReason).WithFailedPlugin("TestPlugin"),
```
Comment (chendave) [inline pkg/scheduler/framework/runtime/framework_test.go:1712]: this is a behavior change, it's changing the status from `Error` to `UnschedulableAndUnresolvable`. This needs to be discussed.

**Review by chendave — COMMENTED** [review]
Just a quickly go through, I will find time to revisit this later if this is still open.

File: pkg/scheduler/framework/runtime/framework_test.go (hunk 3 above)
Comment (kerthcet) [reply pkg/scheduler/framework/runtime/framework_test.go:1712]: Yes, discussion is always needed. 

Currently, after we run `RunBindPlugins` and `RunPreBindPlugins`, we will wrap the status and always return a Error one, then even the pod is rejected for unschedulable reason, we'll also put it back to backingOffQ and activeQ rather than unschedulablePods. See [`MoveAllToActiveOrBackoffQueue`](https://github.com/kubernetes/kubernetes/blob/5debf63a742fab8f3d9b3672a1fd3f14d2289dce/pkg/scheduler/internal/queue/scheduling_queue.go#L626-L636), but it will fail again. So we should distinguish these status apart.

Comment (kerthcet) [reply pkg/scheduler/framework/runtime/framework_test.go:1712]: cc @ahg-g @alculquicondor @Huang-Wei

File: pkg/scheduler/framework/interface.go (hunk 1 above)
Comment (kerthcet) [reply pkg/scheduler/framework/interface.go:253]: https://github.com/kubernetes/kubernetes/blob/5debf63a742fab8f3d9b3672a1fd3f14d2289dce/pkg/scheduler/schedule_one.go#L785-L788

We used to return an error here but now I changed to *framework.Status, like framework.AsStatus(err), but the problem here is err might be nil, so I added these codes. But I think the change makes sense and has no side effects.

File: pkg/scheduler/framework/runtime/framework_test.go (hunk 3 above)
Comment (alculquicondor) [reply pkg/scheduler/framework/runtime/framework_test.go:1712]: The only in-tree plugin that we have uses Error, so I'm fine with this change.

File: pkg/scheduler/framework/runtime/framework.go [hunk 4]
```diff
@@ -1003,6 +1003,11 @@ func (f *frameworkImpl) RunPreBindPlugins(ctx context.Context, state *framework.
 	for _, pl := range f.preBindPlugins {
 		status = f.runPreBindPlugin(ctx, pl, state, pod, nodeName)
 		if !status.IsSuccess() {
+			if status.IsUnschedulable() {
+				klog.V(4).InfoS("Pod rejected by PreBind plugin", "pod", klog.KObj(pod), "plugin", pl.Name(), "status", status.Message())
```
Comment (alculquicondor) [inline pkg/scheduler/framework/runtime/framework.go:1007]: why 4? is there precedent?

File: pkg/scheduler/schedule_one.go [hunk 5]
```diff
@@ -239,9 +239,13 @@ func (sched *Scheduler) bindingCycle(ctx context.Context, state *framework.Cycle
 			// Avoid moving the assumed Pod itself as it's always Unschedulable.
 			// It's intentional to "defer" this operation; otherwise MoveAllToActiveOrBackoffQueue() would
 			// update `q.moveRequest` and thus move the assumed pod to backoffQ anyways.
-			defer sched.SchedulingQueue.MoveAllToActiveOrBackoffQueue(internalqueue.AssignedPodDelete, func(pod *v1.Pod) bool {
-				return assumedPod.UID != pod.UID
-			})
+			if waitOnPermitStatus.IsUnschedulable() {
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one.go:242]: why this change?

File: pkg/scheduler/framework/runtime/framework.go (hunk 4 above)
Comment (kerthcet) [reply pkg/scheduler/framework/runtime/framework.go:1007]: Yes, see https://github.com/kubernetes/kubernetes/blob/863a462eeccca4c24bf4c4e9012f5ebccade70a2/pkg/scheduler/framework/runtime/framework.go#L1150-L1154

File: pkg/scheduler/schedule_one.go (hunk 5 above)
Comment (kerthcet) [reply pkg/scheduler/schedule_one.go:242]: For unschedulable pods, since it's useless to put it back to the backingoffQ/activeQ at that moment, so we should sent it to the `unschedulingPods`. Considering `MoveAllToActiveOrBackoffQueue()` would update `q.moveRequest` then `FailureHandler` will move the pod to backoffQ anyways, so we defer the move on purpose.

Comment (alculquicondor) [reply pkg/scheduler/schedule_one.go:242]: So, on the other hand, if there is a failure, you want the pod to be back into backoff/active immediately, right?

@Huang-Wei wdyt?

File: pkg/scheduler/schedule_one.go [hunk 6]
```diff
@@ -259,14 +263,20 @@ func (sched *Scheduler) bindingCycle(ctx context.Context, state *framework.Cycle
 			// "Forget"ing an assumed Pod in binding cycle should be treated as a PodDelete event,
 			// as the assumed Pod had occupied a certain amount of resources in scheduler cache.
 			// TODO(#103853): de-duplicate the logic.
-			sched.SchedulingQueue.MoveAllToActiveOrBackoffQueue(internalqueue.AssignedPodDelete, nil)
+			if preBindStatus.IsUnschedulable() {
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one.go:266]: @Huang-Wei same question.

File: pkg/scheduler/schedule_one.go [hunk 7]
```diff
@@ -775,23 +791,17 @@ func (sched *Scheduler) assume(assumed *v1.Pod, host string) error {
...
-	if bindStatus.IsSuccess() {
-		return nil
-	}
-	if bindStatus.Code() == framework.Error {
-		return bindStatus.AsError()
+		return framework.AsStatus(err)
 	}
-	return fmt.Errorf("bind status: %s, %v", bindStatus.Code().String(), bindStatus.Message())
+	status = fwk.RunBindPlugins(ctx, state, assumed, targetNode)
+	return
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one.go:804]:
```suggestion
	return fwk.RunBindPlugins(ctx, state, assumed, targetNode)
```

File: pkg/scheduler/schedule_one.go (hunk 5 above)
Comment (kerthcet) [reply pkg/scheduler/schedule_one.go:242]: This is what `PreBind` and `Bind` phases behave like now. And WaitOnPermit here will always return a Unschedulable status when rejected, see https://github.com/kubernetes/kubernetes/blob/c870f1ddc73903c074b441a11fe16dff23fc1438/pkg/scheduler/framework/runtime/waiting_pods_map.go#L152-L165

I think we can unify the behaviors when getting different status, this is also what I want to achieve in this PR. Like 
- unschedulable -> unschedulablePods
- error -> backoffQ

Do we have any special cases ?

Comment (Huang-Wei) [reply pkg/scheduler/schedule_one.go:242]: > unschedulable -> unschedulablePods
> error -> backoffQ

The semantics LGTM.

It's just w/ this change, we start to honor the status returned by PreBind/Bind. For waitForPermit, a bit over-engineering though as it's not a public interface and in all cases it's treated as Unschedulable.

@cofyc FYI: for now, VolumeBind plugin's PreBind returns either Error or Success. In terms of Error, actually scheduler will retry them, is that intentional? If not, say you have already registered  a cluster event in EventsToRegister(), we can change Error to Unschedulable, so scheduler will hold rescheduling it until a related event is fired.

File: pkg/scheduler/framework/runtime/framework.go [hunk 8]
```diff
@@ -1030,15 +1035,20 @@ func (f *frameworkImpl) RunBindPlugins(ctx context.Context, state *framework.Cyc
...
-	for _, bp := range f.bindPlugins {
-		status = f.runBindPlugin(ctx, bp, state, pod, nodeName)
+	for _, pl := range f.bindPlugins {
+		status = f.runBindPlugin(ctx, pl, state, pod, nodeName)
 		if status.IsSkip() {
 			continue
 		}
 		if !status.IsSuccess() {
+			if status.IsUnschedulable() {
+				klog.V(4).InfoS("Pod rejected by Bind plugin", "pod", klog.KObj(pod), "plugin", pl.Name(), "status", status.Message())
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/runtime/framework.go:1045]: remove empty lines. (applies elsewhere)

File: pkg/scheduler/schedule_one.go [hunk 9]
```diff
@@ -239,9 +239,13 @@ func (sched *Scheduler) bindingCycle(ctx context.Context, state *framework.Cycle
...
 			// update `q.moveRequest` and thus move the assumed pod to backoffQ anyways.
-			defer sched.SchedulingQueue.MoveAllToActiveOrBackoffQueue(internalqueue.AssignedPodDelete, func(pod *v1.Pod) bool {
-				return assumedPod.UID != pod.UID
-			})
+			if waitOnPermitStatus.IsUnschedulable() {
+				defer sched.SchedulingQueue.MoveAllToActiveOrBackoffQueue(internalqueue.AssignedPodDelete, func(pod *v1.Pod) bool {
+					return assumedPod.UID != pod.UID
+				})
+			} else {
+				sched.SchedulingQueue.MoveAllToActiveOrBackoffQueue(internalqueue.AssignedPodDelete, nil)
```
Comment (Huang-Wei) [inline pkg/scheduler/schedule_one.go:247]: remove empty lines. (applies elsewhere)

File: pkg/scheduler/schedule_one.go (hunk 5 above)
Comment (kerthcet) [reply pkg/scheduler/schedule_one.go:242]: > It's just w/ this change, we start to honor the status returned by PreBind/Bind. For waitForPermit, a bit over-engineering though as it's not a public interface and in all cases it's treated as Unschedulable.

The intention here is to unify the logic then in this PR https://github.com/kubernetes/kubernetes/pull/112025, I can extract them and reduce code redundant.

File: pkg/scheduler/framework/runtime/framework.go (hunk 8 above)
Comment (kerthcet) [reply pkg/scheduler/framework/runtime/framework.go:1045]: What do you mean `empty lines`? We do not have here.

**Comment by kerthcet** [comment]
Rebased

File: pkg/scheduler/framework/runtime/framework_test.go (hunk 3 above)
Comment (ahg-g) [reply pkg/scheduler/framework/runtime/framework_test.go:1738]: This is fine even if bind was returning unschedulable.

File: pkg/scheduler/schedule_one.go (hunk 5 above)
Comment (ahg-g) [reply pkg/scheduler/schedule_one.go:242]: While the semantics make sense, the nesting of the logic here is a future mine field; consider simplifying this somehow, perhaps converting this into a switch statement with cases?

File: pkg/scheduler/framework/runtime/framework.go (hunk 8 above)
Comment (Huang-Wei) [reply pkg/scheduler/framework/runtime/framework.go:1045]: weird... I did see empty lines on the PR view yesterday. Nvm then.

File: pkg/scheduler/framework/runtime/framework_test.go (hunk 3 above)
Comment (Huang-Wei) [reply pkg/scheduler/framework/runtime/framework_test.go:1738]: Fine with this change to honor the status returned by plugins. Also tagged @cofyc above to see he wants to change Error to Unschedulable so we don't immediately retry pods that failed on PreBind of volumebinding. (w/o this PR, no matter what status is returned from PreBind/Bind, it's treated as Error)

File: pkg/scheduler/schedule_one.go (hunk 5 above)
Comment (Huang-Wei) [reply pkg/scheduler/schedule_one.go:242]: I think that's the intention. @kerthcet mentioned in the description section, that he will leave the real refactoring in the follow-up.

**Comment by Huang-Wei** [comment]
LGTM. @kerthcet mind squshing?

File: pkg/scheduler/schedule_one.go (hunk 5 above)
Comment (kerthcet) [reply pkg/scheduler/schedule_one.go:242]: Yeah, I'll leave the refactor to that PR.

**Comment by kerthcet** [comment]
> LGTM. @kerthcet mind squshing?

Squashed.

File: test/integration/scheduler/plugins/plugins_test.go [hunk 10]
```diff
@@ -343,12 +343,12 @@ func (pp *PreBindPlugin) PreBind(ctx context.Context, state *framework.CycleStat
...
 	pp.podUIDs[pod.UID] = struct{}{}
-	if pp.failPreBind {
-		return framework.NewStatus(framework.Error, fmt.Sprintf("injecting failure for pod %v", pod.Name))
-	}
 	if pp.rejectPreBind {
 		return framework.NewStatus(framework.Unschedulable, fmt.Sprintf("reject pod %v", pod.Name))
 	}
+	if pp.failPreBind {
+		return framework.NewStatus(framework.Error, fmt.Sprintf("injecting failure for pod %v", pod.Name))
+	}
```
Comment (chendave) [inline test/integration/scheduler/plugins/plugins_test.go:351]: I understand the intention for this change, but this is not necessary, I am fine with it though.

File: pkg/scheduler/framework/runtime/framework.go (hunk 4 above)
Comment (chendave) [reply pkg/scheduler/framework/runtime/framework.go:1007]: "Pod rejected by PreBind plugin" looks weird in the context of `prebind`, this is should be different with `permit` pulgin.

How about "Pod is unschedulable with node...."  or something similar? also pls pass in the node name to reveal more information.

**Review by chendave — COMMENTED** [review]
Nice job!

File: test/integration/scheduler/plugins/plugins_test.go (hunk 10 above)
Comment (kerthcet) [reply test/integration/scheduler/plugins/plugins_test.go]: Revert.

File: pkg/scheduler/framework/runtime/framework.go (hunk 4 above)
Comment (kerthcet) [reply pkg/scheduler/framework/runtime/framework.go]: > "Pod rejected by PreBind plugin" looks weird in the context of `prebind`, this is should be different with `permit` pulgin.
> 
I think the word is appropriate, see annotation of PreBind: https://github.com/kubernetes/kubernetes/blob/bcea98234f0fdc5ce4e976758977eec72b13bfd4/pkg/scheduler/framework/interface.go#L461-L463


> also pls pass in the node name to reveal more information.

Make sense.

**Comment by alculquicondor** [comment]
/hold
I would like to give a review too. As these are critical code paths, I think it's better to have more eyes on it.

File: test/integration/scheduler/plugins/plugins_test.go [hunk 11]
```diff
@@ -1022,14 +1016,18 @@ func TestPrebindPlugin(t *testing.T) {
...
+			if test.fail {
 				if test.succeedOnRetry {
 					if err = testutils.WaitForPodToScheduleWithTimeout(testCtx.ClientSet, pod, 10*time.Second); err != nil {
 						t.Errorf("Expected the pod to be schedulable on retry, but got an error: %v", err)
 					}
 				} else if err = wait.Poll(10*time.Millisecond, 30*time.Second, podSchedulingError(testCtx.ClientSet, pod.Namespace, pod.Name)); err != nil {
 					t.Errorf("Expected a scheduling error, but didn't get it. error: %v", err)
 				}
+			} else if test.reject {
+				if err = testutils.WaitForPodToScheduleWithTimeout(testCtx.ClientSet, pod, 5*time.Second); err == nil {
```
Comment (alculquicondor) [inline test/integration/scheduler/plugins/plugins_test.go:1028]: Can you use WaitForPodUnschedulable instead?

Comment (kerthcet) [reply test/integration/scheduler/plugins/plugins_test.go:1028]: `WaitForPodUnschedulable` also call `WaitForPodToScheduleWithTimeout` underlying, and the timeout will be 30s, it's too long I think.

Comment (alculquicondor) [reply test/integration/scheduler/plugins/plugins_test.go:1028]: it calls `WaitForPodUnschedulableWithTimeout`, which relies on the Pod condition to show up in the Pod status.

Comment (kerthcet) [reply test/integration/scheduler/plugins/plugins_test.go]: Updated

**Comment by kerthcet** [comment]
Failed in unit test here https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/112103/pull-kubernetes-unit/1567709021257863168, but I couldn't reproduce it. And in my local env, I haven't met this before. 

And the test failure seems not related to this PR.

**Comment by alculquicondor** [comment]
Is there an open issue for that?

File: test/integration/scheduler/plugins/plugins_test.go [hunk 12]
```diff
@@ -1341,7 +1339,8 @@ func TestUnReservePreBindPlugins(t *testing.T) {
 			}
 
 			if test.fail {
```
Comment (alculquicondor) [inline test/integration/scheduler/plugins/plugins_test.go:1341]: should it be `test.wantUnschedulable` instead?

File: test/integration/scheduler/plugins/plugins_test.go [hunk 13]
```diff
@@ -1341,7 +1339,8 @@ func TestUnReservePreBindPlugins(t *testing.T) {
...
-				if err = wait.Poll(10*time.Millisecond, 30*time.Second, podSchedulingError(testCtx.ClientSet, pod.Namespace, pod.Name)); err != nil {
+				if err = waitForPodUnschedulable(testCtx.ClientSet, pod); err != nil {
+					// if err = wait.Poll(10*time.Millisecond, 30*time.Second, podSchedulingError(testCtx.ClientSet, pod.Namespace, pod.Name)); err != nil {
```
Comment (alculquicondor) [inline test/integration/scheduler/plugins/plugins_test.go:1343]: remove this line?

File: test/integration/scheduler/plugins/plugins_test.go (hunk 12 above)
Comment (kerthcet) [reply test/integration/scheduler/plugins/plugins_test.go:1341]: Change to `test.reject` which corresponding to `PreBindPlugin.rejectPreBind`.

File: test/integration/scheduler/plugins/plugins_test.go (hunk 13 above)
Comment (kerthcet) [reply test/integration/scheduler/plugins/plugins_test.go:1343]: done

**Comment by kerthcet** [comment]
> Is there an open issue for that?

It's actually a issue created by you https://github.com/kubernetes/kubernetes/issues/107414 🤔

File: test/integration/scheduler/plugins/plugins_test.go [hunk 14]
```diff
@@ -1338,9 +1338,8 @@ func TestUnReservePreBindPlugins(t *testing.T) {
 				t.Errorf("Error while creating a test pod: %v", err)
 			}
 
-			if test.fail {
+			if test.reject {
```
Comment (alculquicondor) [inline test/integration/scheduler/plugins/plugins_test.go:1341]: it should have been `wantReject`, but that's a nit.

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve
/hold
for the nit

File: test/integration/scheduler/plugins/plugins_test.go (hunk 14 above)
Comment (kerthcet) [reply test/integration/scheduler/plugins/plugins_test.go]: done
